### ISSUE TITLE
Redundant tile loads in MBXRasterTileRenderer canDrawMapRect: when using 512px tiles

### ISSUE THREAD

**Issue by s219** [mapbox/mbxmapkit#167]
I stumbled on a serious bug, one that unnecessarily increases the number of tile loads by as much as 3-4X when using 512px tiles.  This not only has a big impact on performance, but also how we would potentially get billed for tile downloads.

MBXRasterTileRenderer's ``canDrawMapRect:`` receives requests to draw 256px mapRects.  On a retina device, a parent 512px tile path is created, which would encompass the child 256px mapRect.  The 512px tile is requested to load, but the active download queue is managed in terms of the original 256px mapRect path (an NSString called ``xyz`` formed from the path's x,y,z identifies each download).  What that means is that if other 256px mapRects come along needing display, they will request their own 512px tiles to load, even if those 256px mapRects would be a child to the very same 512px parent tile that is already in the active download queue.  

The end result is a lot of redundant tile loads on a retina device, a factor of 3 to 4 too many.  If the mapView wants to draw the four 256px mapRects in a 512px tile area, that 512px tile will be requested a total of four times even though it's just needed once!  I first noticed this on a slow network connection when testing cache strategies -- I was burning through tile downloads, requesting 28-31 512px tiles to fill the screen on an iPhone 6 which should only need 9-12 tiles, tops.  When revisiting a cached region on the map, this doesn't pose a huge issue, but for every new region, MBXMapKit is going to be requesting 3-4X too many redundant tiles from the MapBox tile servers before in-app caching with ``NSURLCache`` starts to help out.  Obviously that's not good for billing, but it also hurts performance too.

The quick fix here is to manage the active tile download queue in terms of the tile path id string of the 512px tile, such that follow-on requests that can use the same 512px tile don't request it once again.  Doing that cut my per-screen tile usage down from 28 to 9 for the default map loaded when my app launches, which is a big difference.

The side effect of managing downloads by the 512px tile path id string is that we're still getting requests to draw 256px mapRects, and we need to deal with those.  The current coding setup of ``canDrawMapRect:`` does that just fine, however it results in thousands of deferred ``setNeedsDisplayInMapRect:`` calls for 256px mapRects that are waiting on a 512px tile to show up.  This in turn triggers the very same crash that has been haunting some of us, discussed in #163 and #153.

What I am testing now is canceling all incoming requests for 256px mapRects except for the ones that spawned active downloads of a 512px tile.  When those downloads finish, they fire a completion block where I now create a new mapRect for the entire 512px tile's rect, and call ``setNeedsDisplayInMapRect:`` on it.  I am playing around to see how performance scales with breaking that up into smaller rects instead of a single 512px rect.

So far this is working marvelously and I am seeing better performance and far fewer tile downloads.  I want to test this more before posting any code, and I really need to dig out one of my old non-retina devices to make sure I haven't broke anything when dealing with straight 256px tiles (in that case, I am pretty sure my approach falls back to the old strategy).  But in the meantime, I wanted to give everyone a heads up, as this bug could seriously be inflating your tile download numbers and billed cost.

**Comment by s219** [comment]
Here is my revised ``canDrawMapRect`` that fixes this issue and has been working well in testing.  It also incorporates fixes to avoid a crash discussed in #163 and #153.  This version of ``canDrawMapRect`` does not use the drawing cycle as a delayed looping mechanism -- it only calls ``setNeedsDisplayInMapRect`` when there is actually a mapRect to draw.


```
- (BOOL)canDrawMapRect:(MKMapRect)mapRect zoomScale:(MKZoomScale)zoomScale {
    
    MKTileOverlay *tileOverlay = (MKTileOverlay *)self.overlay;
    
    MKTileOverlayPath path = [self pathForMapRect:mapRect zoomScale:zoomScale];
    
    BOOL usingBigTiles = (tileOverlay.tileSize.width == 512);
    
    MKTileOverlayPath childPath = path;
    
    if (usingBigTiles) {
        path.x /= 2;
        path.y /= 2;
        path.z -= 1;
    }
    
    NSString *xyz = [[self class] xyzForPath:childPath];
    
    NSString *xyzQueue = [[self class] xyzForPath:path];
    
    BOOL tileReady = NO;
    
    // introduce a new tileRect that covers the entire region of a 512px tile
    
    MKMapRect tileRect;
    
    if (usingBigTiles) {
        
        double xTile = 256.0*path.x/(0.5*zoomScale);
        double yTile = 256.0*path.y/(0.5*zoomScale);
        double wTile = 2.0*mapRect.size.width;
        
        tileRect = MKMapRectMake(xTile, yTile, wTile, wTile);
        
    } else {
        
        tileRect=mapRect;
        
    }
        
    @synchronized(self) {
        tileReady = ([[self class] imageDataFromRenderer:self forXYZ:xyz usingBigTiles:usingBigTiles] != nil);
    }

    if (tileReady) {

        return YES;
        
    } else {
        
        __weak typeof(self) weakSelf = self;
        
        BOOL tileActive = NO;
        @synchronized(weakSelf) {
            tileActive = ([weakSelf.activeDownloads containsObject:xyzQueue]);
            if (!tileActive) {
                [weakSelf.activeDownloads addObject:xyzQueue];
            }
        }
        
        if (!tileActive) {
            
            [(MKTileOverlay *)weakSelf.overlay loadTileAtPath:path result:^(NSData *tileData, NSError *error) {
                
                @synchronized(weakSelf) {
                    [weakSelf.activeDownloads removeObject:xyzQueue];
                }
                
                if (tileData) {
                    
                    NSData *tileDataCopy = [[NSData alloc] initWithBytes:tileData.bytes length:tileData.length];
                    
                    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                        
                        CGDataProviderRef provider = CGDataProviderCreateWithCFData((CFDataRef)tileDataCopy);
                        CGImageRef imageRef = nil;
                        
                        if ([[weakSelf class] dataIsPNG:tileDataCopy]) {
                            imageRef = CGImageCreateWithPNGDataProvider(provider, nil, NO, kCGRenderingIntentDefault);
                        } else if ([[weakSelf class] dataIsJPEG:tileDataCopy]) {
                            imageRef = CGImageCreateWithJPEGDataProvider(provider, nil, NO, kCGRenderingIntentDefault);
                        }
                        
                        if (imageRef) {
                            @synchronized(weakSelf) {
                                [[weakSelf class] addImageData:tileDataCopy
                                                    toRenderer:weakSelf
                                                        forXYZ:xyz
                                                 usingBigTiles:usingBigTiles];
                            }
                        }
                        
                        CGImageRelease(imageRef);
                        CGDataProviderRelease(provider);

                        // note: now calling setNeedsDisplay from the same thread that adds image data to renderer 

                        [weakSelf setNeedsDisplayInMapRect:tileRect zoomScale:zoomScale]; 
                                              
                    });
                    
                }
                
            }];
            
        }
        
        return NO;
        
    }
    
}
```

The other somewhat important change is to move ``setNeedsDisplay`` into the block of the thread that adds image data to the renderer, so that ``setNeedsDisplay`` only gets called **after** the new tile's data is available and ready to go.  Previously, I was seeing cases where an extra download would be fired off for some tiles because they were removed from the active download queue (thus triggering ``!tileActive``) before they became ready for the renderer (to trigger ``tileReady``).

All together, this fix reduces the number of new 512px tiles needed for each new screenful of mapping by a factor of 3-4, which is huge.

Final note: the ``setNeedsDisplay`` call marks the entire tile's 512px region as needing update, but that automatically gets broken down into four smaller 256px mapRects before going to drawRect.  So there is no need to break it down in our code.

**Comment by friedbunny** [comment]
Thanks for tracking this down, @s219! I'll look at your code and do stability testing before putting this into a real pull request.

**Comment by friedbunny** [comment]
Still seeing VectorKit crashes, but they appear more rare.

```c
frame #1: 0x00000001119787a7 VectorKit`-[VKTileKeyMap objectForKey:] + 23
frame #2: 0x00000001119fb5bc VectorKit`-[VKTilePool objectForKey:] + 156
frame #3: 0x0000000111720bab VectorKit`-[VKTile setPool:] + 59
frame #4: 0x0000000111720c8b VectorKit`-[VKTile dealloc] + 27
frame #5: 0x000000011177dee6 VectorKit`-[VKGGLRasterOverlayTile dealloc] + 86
frame #6: 0x000000010dc9328e libobjc.A.dylib`objc_object::sidetable_release(bool) + 236
frame #7: 0x0000000111b43e33 VectorKit`__40-[VKRasterOverlayTileSource _queueDraw:]_block_invoke + 755
frame #8: 0x000000010e75e186 libdispatch.dylib`_dispatch_call_block_and_release + 12
frame #9: 0x000000010e77d614 libdispatch.dylib`_dispatch_client_callout + 8
frame #10: 0x000000010e767552 libdispatch.dylib`_dispatch_root_queue_drain + 1768
frame #11: 0x000000010e768b17 libdispatch.dylib`_dispatch_worker_thread3 + 111
frame #12: 0x000000010eaea637 libsystem_pthread.dylib`_pthread_wqthread + 729
frame #13: 0x000000010eae840d libsystem_pthread.dylib`start_wqthread + 13
```

**Comment by s219** [comment]
@friedbunny , are you doing anything special to flush out VectorKit crashes, or just general use of the map?

From what I could see, the VectorKit crashes were very proportional to the number of times we cycle through the drawing loop.  So the crashes were very obvious (and usually happened within 5 minutes) in previous situations where absent offline tiles triggered an endless loop, or where online tile loading deferred drawing with repeated trips through canDrawMapRect.

If we can come up with any info to pin down the VectorKit crashes, I'd be glad to burn a tech support incident with Apple to see if there is a workaround, or at least to get a possible bug in front of the right person at Apple.  If there is a repeatable way to trigger the crashes, it would also be good to try on iOS 9 beta to see if there is any difference.

Thanks!

**Comment by friedbunny** [comment]
It took me quite a while to trigger one — had to pan and zoom around an online map for ~10 minutes. I'm working on finding a simpler/faster way to reproduce, clearly 10 minutes is not tenable. ;)

**Comment by joteichmann** [comment]
Phantastic! The map load 4x faster!!! Awesome work, credits to @s219
